runtime/jvmti/jvmtiCapability.c

```diff
@@ -302,126 +302,97 @@ jvmtiAddCapabilities(jvmtiEnv* env,
 	jvmtiCapabilities newCapabilities;
 	UDATA i;
 	jvmtiError rc = JVMTI_ERROR_NOT_AVAILABLE;
+	J9VMThread *currentThread = NULL;
 
 	Trc_JVMTI_jvmtiAddCapabilities_Entry(env);
 
 	ENSURE_PHASE_ONLOAD_OR_LIVE(env);
 
 	ENSURE_NON_NULL(capabilities_ptr);
 
-	/* Handle can_tag_objects immediately as the requiredDebugAttributes check requires exclusive VM access
-	 * which must not occur while holding the JVMTI mutex.
-	 */
-	if (capabilities_ptr->can_tag_objects) {
-		J9InternalVMFunctions const * const vmFuncs = vm->internalVMFunctions;
-		J9HookInterface ** vmHook = vmFuncs->getVMHookInterface(vm);
-		/* If J9HOOK_VM_REQUIRED_DEBUG_ATTRIBUTES has not been disabled yet (i.e. VM is starting up),
-		 * there's no need to handle can_tag_objects specially.
-		 */
-		if (0 == (*vmHook)->J9HookIsEnabled(vmHook, J9HOOK_VM_REQUIRED_DEBUG_ATTRIBUTES)) {
-			J9VMThread *currentThread = NULL;
-			jvmtiError tempRC = getCurrentVMThread(vm, &currentThread);
-			if (tempRC != JVMTI_ERROR_NONE) {
-				rc = tempRC;
-				goto fail;
-			}
-			vmFuncs->internalEnterVMFromJNI(currentThread);
-			vmFuncs->acquireExclusiveVMAccess(currentThread);
-			if (0 == (vm->requiredDebugAttributes & J9VM_DEBUG_ATTRIBUTE_ALLOW_USER_HEAP_WALK)) {
-				J9MemoryManagerFunctions const * const mmFuncs = vm->memoryManagerFunctions;
-				vm->requiredDebugAttributes |= J9VM_DEBUG_ATTRIBUTE_ALLOW_USER_HEAP_WALK;
-				/* J9MMCONSTANT_EXPLICIT_GC_RASDUMP_COMPACT allows the GC to run while the current thread is holding
-				 * exclusive VM access.
-				 */
-				mmFuncs->j9gc_modron_global_collect_with_overrides(currentThread, J9MMCONSTANT_EXPLICIT_GC_RASDUMP_COMPACT);
-				if (J9_GC_POLICY_METRONOME == vm->gcPolicy) {
-					/* In metronome, the previous GC call may have only finished the current cycle.
-					 * Call again to ensure a full GC takes place.
-					 */
-					mmFuncs->j9gc_modron_global_collect_with_overrides(currentThread, J9MMCONSTANT_EXPLICIT_GC_RASDUMP_COMPACT);
-				}
-			}
-			vmFuncs->releaseExclusiveVMAccess(currentThread);
-			vmFuncs->internalExitVMToJNI(currentThread);
-		}
-	}
+	rc = getCurrentVMThread(vm, &currentThread);
+	if (rc == JVMTI_ERROR_NONE) {
+		rc = JVMTI_ERROR_NOT_AVAILABLE;
+		vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);
```

Comment:
Does this need vm access?

Reply:
VM access was added here as a lock for requiredDebugAttributes - it's read/written by the ensureHeapWalkable code under exclusive. There are other places in the code which update this field without any protection, but they're all startup-related.